### ISSUE TITLE
Net outs collapse to zero values in continuous phase on 2D experiments?

### ISSUE THREAD

**Issue by yeshwanth95** [czq142857/BSP-NET-pytorch#9]
Hi there!

I have been experimenting with the 2D version of your BSP_SVR model in a PyTorch implementation of mine. I have been able to reproduce the results on the toy dataset. However, when I switched to more complex binary masks (such as segmentation pipeline outputs), I'm noticing that the bsp_loss_sp loss initially converges and then after a few epochs, collapses to a constant value in the continuous phase. When this happens, I notice that the net_outs tend to collapse to a tensor of all zero values.
However, initializing the model on these weights and continuing to train in the discrete phase still seems to work in the expected direction, albeit, the convergence of losses is very slow.

I've noticed that the model only works well with a very low learning rate (0.00002) in the 2D case. However, it appears the model takes a very long time to train using this learning rate on more complex binary masks. Do you have any suggestions or insights as to why the model collapses to all zero values (in the 2D case where binary masks are more complex than the toy dataset)? And on how this could be tackled?

Looking forward to your reply. Thank you.

**Comment by czq142857** [comment]
Hi,

I am not familiar with your experimental setup, and I have never experienced such collapses, therefore my comments might not be very helpful.

I suspect the collapses might have something to do with the gradients. There are many ReLU and clipping operations in the shallow MLP, therefore once the network outputs zeros, it will receive no gradients and will be stuck there forever. I guess sometimes a specific batch could bring abnormally strong gradients and make the network stuck instantly, and this is totally random. I am not sure how this could be fixed.

**Comment by yeshwanth95** [comment]
Hi @czq142857. Thanks for the reply!

Actually, you are right about the network being stuck after it outputs zeros. The gradient propagation stops entirely at this point. Although this happens pretty much around the same epoch in each run. So it could be due to a specific 'difficult' batch that causes this, given that the batch ordering is not randomized.
I too think the many clamping operations could cause this behaviour. I'll have to take a closer look though.

**Comment by yeshwanth95** [comment]
Hi @czq142857. A few quick questions.

1. I've just been wondering if there is any reason you used a learning rate of 0.00002 (quite a low value) in the 2D experiments with the toy dataset? It appears that the above-mentioned behaviour only occurs when I try to increase the learning rate above this value with the toy dataset.

2. Also, when I use more complex binary masks, the learning rate needs to be lowered even further. When I lower the learning rate, I'm able to bypass the issue of outputs collapsing to zero however at the cost of the model converging extremely slowly. Are there any limitations to BSPNet I'm overlooking that need to be addressed so that it generalises well to more complex shapes?

I've attached some examples of the binary masks I'm trying to train BSPNet on. Would love to hear your thoughts!
![0_gt](https://user-images.githubusercontent.com/54409589/155748000-6b430404-ac05-41ba-aaec-356b32805d37.png)
![1_gt](https://user-images.githubusercontent.com/54409589/155748002-2fc0e65b-84bf-4836-a360-4ce2b1f64543.png)
![2_gt](https://user-images.githubusercontent.com/54409589/155748005-cee68076-f08e-4a6a-a676-c54b97aefcdd.png)
![3_gt](https://user-images.githubusercontent.com/54409589/155748007-7b507d54-3a04-462a-ac97-6f2c9298acf6.png)
![4_gt](https://user-images.githubusercontent.com/54409589/155748009-7795158c-029a-4efa-b352-ab0899f235a1.png)

**Comment by czq142857** [comment]
1. I cannot remember why I was setting the learning rate to 0.00002. I tried 0.0001 just now and it also worked. But the loss fluctuated between 0.001 and 0.005 at the end, so I guess I used a small learning rate to avoid such fluctuation. Nevertheless, it should not cause the collapse. Please try the original implementation and see if there is something different in your pytorch implementation,
2. BSP-Net requires the shapes in the dataset to have some sort of correspondences (see Figure 4 in the paper). Each part (convex) is shared by different shapes, and ideally it should represent the same or corresponding parts in those different shapes. I do not see such correspondences in your dataset, therefore I do not think BSP-Net would work, unless you try to overfit on a handful of samples.

**Comment by yeshwanth95** [comment]
Hi @czq142857! Thanks for the clarifications. I'll check again to see if there are any differences between my implementation and the original.

Although, I think you are right about the dataset. I'm running some experiments with a simpler version of this dataset (smaller patches basically) and BSP-Net seems to perform better than the earlier runs. However, I still feel my dataset also has correspondences between parts across different objects (since it's just gt masks of mostly rectangular buildings in airborne images). I'll try a few more experiments with the same dataset on the original tf1 implementation to make sure this is not an implementation issue.
